drivers/can/can_numaker.c

```diff
@@ -0,0 +1,316 @@
+
+	LOG_DBG("Clock source/divider: %d/%d", clksrc_rate, clkdiv_divider);
+	*rate = clksrc_rate / clkdiv_divider;
+
+cleanup:
+
+	return rc;
+}
+
+static int can_numaker_init(const struct device *dev)
+{
+	const struct can_mcan_config *mcan_config = dev->config;
+	const struct can_numaker_config *config = mcan_config->custom;
+	struct numaker_scc_subsys scc_subsys;
+	int rc = 0;
+	uint32_t rate;
+
+	/* Validate this module's reset object */
+	if (!device_is_ready(config->reset.dev)) {
+		LOG_ERR("reset controller not ready");
+		return -ENODEV;
+	}
+
+	SYS_UnlockReg();
+
+	memset(&scc_subsys, 0x00, sizeof(scc_subsys));
+	scc_subsys.subsys_id        = NUMAKER_SCC_SUBSYS_ID_PCC;
+	scc_subsys.pcc.clk_modidx   = config->clk_modidx;
+	scc_subsys.pcc.clk_src      = config->clk_src;
+	scc_subsys.pcc.clk_div      = config->clk_div;
+
+	/* Equivalent to CLK_EnableModuleClock() */
+	rc = clock_control_on(config->clk_dev, (clock_control_subsys_t) &scc_subsys);
+	if (rc < 0) {
+		goto cleanup;
+	}
+	/* Equivalent to CLK_SetModuleClock() */
+	rc = clock_control_configure(config->clk_dev, (clock_control_subsys_t)&scc_subsys, NULL);
+	if (rc < 0) {
+		goto cleanup;
+	}
+
+	/* Configure pinmux (NuMaker's SYS MFP) */
+	rc = pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+	if (rc < 0) {
+		goto cleanup;
+	}
+
+	/* Reset CAN to default state, same as BSP's SYS_ResetModule(id_rst) */
+	reset_line_toggle_dt(&config->reset);
+
+	config->irq_config_func(dev);
+
+	rc = can_mcan_configure_mram(dev, config->mrba, config->mram);
+	if (rc != 0) {
+		return rc;
+	}
+
+	rc = can_mcan_init(dev);
+	if (rc < 0) {
+		LOG_ERR("Failed to initialize mcan: %d", rc);
+		goto cleanup;
+	}
+
+	rc = can_numaker_get_core_clock(dev, &rate);
+	if (rc < 0) {
+		goto cleanup;
+	}
+	LOG_INF("CAN core clock: %d", rate);
+
+cleanup:
+
+	SYS_LockReg();
+	return rc;
+}
+
+static const struct can_driver_api can_numaker_driver_api = {
+	.get_capabilities = can_mcan_get_capabilities,
+	.start = can_mcan_start,
+	.stop = can_mcan_stop,
+	.set_mode = can_mcan_set_mode,
+	.set_timing = can_mcan_set_timing,
+	.send = can_mcan_send,
+	.add_rx_filter = can_mcan_add_rx_filter,
+	.remove_rx_filter = can_mcan_remove_rx_filter,
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+	.recover = can_mcan_recover,
+#endif /* CONFIG_CAN_AUTO_BUS_OFF_RECOVERY */
+	.get_state = can_mcan_get_state,
+	.set_state_change_callback = can_mcan_set_state_change_callback,
+	.get_core_clock = can_numaker_get_core_clock,
+	.get_max_filters = can_mcan_get_max_filters,
+	.get_max_bitrate = can_mcan_get_max_bitrate,
+	/* Nominal bit timing limits
+	 *
+	 * NUMAKER MCAN timing limits are specified in the "Nominal bit timing and
+	 * prescaler register (NBTP)" table in the SoC reference manual.
+	 *
+	 * NOTE: The values here are the "physical" timing limits, whereas the register
+	 *       field limits are physical values minus 1 (which is handled by the
+	 *       register assignments in the common MCAN driver code).
+	 */
```

Comment:
This is standard for all Bosch M_CAN IPs. No need for this comment.

Reply:
Yes and removed.